CsCore/PlainNetClassLib/src/Plugins/CsCore/com/csutil/http/apis/OpenAi.cs

```diff
@@ -248,7 +308,81 @@ public class Usage {
         }
 
     }
+    public class VisionGpt {
+
+        public class Line {
+
+            public readonly string role;
+            public readonly object content;
+
+
+            public Line(string role, List<Dictionary<string, object>> content) {
+                this.role = role;
+                this.content = content;
+            }
+            public Line(Role role, List<Dictionary<string, object>> content) {
+                this.role = role.ToString();
+                this.content = content;
+            }
+            [JsonConstructor]
+            public Line(string role, string content) {
+                this.role = role;
+                this.content = content;
+            }
+            public Line(Role role, string content) {
+                this.role = role.ToString();
+                this.content = content;
+            }
+
+        }
 
+        public enum Role { system, user, assistant }
+
+        public class Request {
+
+            /// <summary> See https://beta.openai.com/docs/models/overview </summary>
+            public string model = "gpt-4-vision-preview";
+
+            /// <summary> The maximum number of tokens to generate in the completion.
+            /// The token count of your prompt plus max_tokens cannot exceed the model's context length.
+            /// Most models have a context length of 2048 tokens (except for the newest models, which support 4096). </summary>
+            public int max_tokens { get; set; }
+            public List<Line> messages { get; set; }
+
+            public Request(List<Line> messages, int max_tokens = 4096) {
+                var tokenCountForMessages = JsonWriter.GetWriter(this).Write(messages).Length;
+                if (max_tokens + tokenCountForMessages > 4096) {
+                    max_tokens = 4096 - tokenCountForMessages;
+                }
+                this.messages = messages;
+                this.max_tokens = max_tokens;
+            }
+        }
+
+        public class Response {
```

Comment:
Please add url with details on the different fields here if available

Reply:
Not available.